js/components/Editor.js

```diff
@@ -48,37 +48,53 @@ const Editor = React.createClass({
     this.editor.on("gutterClick", this.onGutterClick);
   },
 
-  onGutterClick(cm, line, gutter, ev) {
-    let info = cm.lineInfo(line);
-    cm.setGutterMarker(
-      line,
-      "breakpoints",
-      info.gutterMarkers ? null : makeMarker()
-    );
-
-    this.props.addBreakpoint({
-      actor: this.props.selectedSource.get("actor"),
-      line: line + 1
-    });
-  },
-
   componentWillReceiveProps(nextProps) {
     const sourceText = nextProps.sourceText;
-    const cursor = this.editor.getCursor();
+    const editor = this.editor;
+    const cursor = editor.getCursor();
+
+    if(this.props.sourceText !== nextProps.sourceText) {
+      if (sourceText.get("loading")) {
+        editor.setValue("Loading...");
+        return;
+      }
+
+      if (sourceText.get("error")) {
+        editor.setValue("Error");
+        console.error(sourceText.get("error"));
+        return;
+      }
 
-    if (sourceText.get("loading")) {
-      this.editor.setValue("Loading...");
-      return;
+      editor.setValue(sourceText.get("text"));
+      editor.setCursor(cursor);
     }
 
-    if (sourceText.get("error")) {
-      this.editor.setValue("Error");
-      console.error(sourceText.get("error"));
-      return;
+    if(this.props.breakpoints !== nextProps.breakpoints) {
+      // TODO: need to make sure the breakpoinst are up-to-date with
+      // the state. It'd be nice to just re-set all of the breakpoints
+      // here, but that might be too slow.
     }
+  },
 
-    this.editor.setValue(sourceText.get("text"));
-    this.editor.setCursor(cursor);
+  onGutterClick(cm, line, gutter, ev) {
+    const currentBp = this.props.breakpoints.find(bp => {
+      return bp.getIn(["location", "line"]) === line + 1;
+    });
```

Comment:
It could be. I don't think every single data access needs to be a query though. If we expect to get an array of things, I think it's fine to operate on that array instead of having to go back through the queries. The queries mainly abstract out where that array comes from, so you can change around your app state.

You could even say that accessing the line from the location field of a breakpoint should be a query. But we can accept some amount of dependency on the state shape to make things more straight-forward.

So I'm not against it, but I'd rather wait and see if we need that in more than 1 place.